### PR TITLE
 feat: update validation for obs['is_primary_data'] to enforce False if uns['spatial']['is_single'] is False

### PR DESCRIPTION
**Pull request by nayib-jose-gloria**
## Reason for Change

- #834 

## Changes

- add _validate_spatial_is_primary_data to _check_spatial_obs
- refactored some unit tests to use a new pytest fixture, validator with a dataset that is spatial and is_single is False (since we were reusing that set-up in a few spots)

### CODE REVIEW

File: cellxgene_schema_cli/cellxgene_schema/validate.py [hunk 1]
```diff
@@ -1000,13 +1000,14 @@ def _validate_obsm(self):
                     self.errors.append(
                         f"Suffix for embedding key in 'adata.obsm' {key} does not match the regex pattern {regex_pattern}."
                     )
-            else:
+            elif key.lower() != "spatial":
```
Comment (nayib-jose-gloria) [inline cellxgene_schema_cli/cellxgene_schema/validate.py:1003]: diff is from https://github.com/chanzuckerberg/single-cell-curation/issues/827 merged into main

File: cellxgene_schema_cli/cellxgene_schema/validate.py [hunk 2]
```diff
@@ -1000,13 +1000,14 @@ def _validate_obsm(self):
...
                 if not re.match(regex_pattern, key):
                     self.errors.append(
                         f"Embedding key in 'adata.obsm' {key} does not match the regex pattern {regex_pattern}."
                     )
                 self.warnings.append(
-                    f"Embedding key in 'adata.obsm' {key} does not start with X_ and thus will not be available in Explorer"
+                    f"Embedding key in 'adata.obsm' {key} is not 'spatial' nor does it start with 'X_'. Thus, it will "
```
Comment (nayib-jose-gloria) [inline cellxgene_schema_cli/cellxgene_schema/validate.py:1009]: diff is from https://github.com/chanzuckerberg/single-cell-curation/issues/827 merged into main

File: cellxgene_schema_cli/cellxgene_schema/validate.py [hunk 3]
```diff
@@ -1037,6 +1038,21 @@ def _validate_obsm(self):
...
+        has_spatial_embedding = "spatial" in self.adata.obsm
+        if is_single and not has_spatial_embedding:
+            self.errors.append(
+                "'spatial' embedding is required in 'adata.obsm' if " "adata.uns['spatial']['is_single'] is True."
+            )
+        elif is_single is None and has_spatial_embedding:
+            self.errors.append(
+                "'spatial' embedding is forbidden in 'adata.obsm' if " "adata.uns['spatial']['is_single'] is not set."
+            )
+
```
Comment (nayib-jose-gloria) [inline cellxgene_schema_cli/cellxgene_schema/validate.py:1055]: diff is from https://github.com/chanzuckerberg/single-cell-curation/issues/827 merged into main

File: cellxgene_schema_cli/tests/fixtures/examples_validate.py [hunk 4]
```diff
@@ -142,7 +142,7 @@
             "HANCESTRO:0575",
             "HsapDv:0000003",
             "donor_1",
-            "nucleus",
+            "na",
```
Comment (nayib-jose-gloria) [inline cellxgene_schema_cli/tests/fixtures/examples_validate.py:145]: diff is from https://github.com/chanzuckerberg/single-cell-curation/issues/827 merged into main

File: cellxgene_schema_cli/tests/test_schema_compliance.py [hunk 5]
```diff
@@ -1993,30 +2010,18 @@ def test_obsm_values_no_X_embedding__non_spatial_dataset(self, validator_with_ad
             "WARNING: Validation of raw layer was not performed due to current errors, try again after fixing current errors.",
         ]
 
-    @pytest.mark.parametrize(
```
Comment (nayib-jose-gloria) [inline cellxgene_schema_cli/tests/test_schema_compliance.py:1996]: split this test up into two for simplicity in the set-up

File: cellxgene_schema_cli/tests/test_schema_compliance.py [hunk 6]
```diff
@@ -2025,24 +2030,9 @@ def test_obsm_values_spatial_embedding_missing__is_single_true(self, validator_w
...
+        validator = validator_with_spatial_and_is_single_false
         del validator.adata.obsm["spatial"]
-        # Format adata.obs into valid shape for Visium and is_single False.
-        validator.adata.obs.pop("array_col")
-        validator.adata.obs.pop("array_row")
-        validator.adata.obs.pop("in_tissue")
-        validator.validate_adata()
-        assert validator.errors == []
-
-    def test_obsm_values_spatial_embedding_present__is_single_false(self, validator_with_visium_assay):
```
Comment (nayib-jose-gloria) [inline cellxgene_schema_cli/tests/test_schema_compliance.py:2039]: moved this test into `test_validate.py` as `test__validate_spatial_is_single_false_ok` and renamed now that we have the fixture `adata_spatial_is_single_false` with spatial embedding present

File: cellxgene_schema_cli/cellxgene_schema/validate.py [hunk 7]
```diff
@@ -1382,6 +1382,25 @@ def _check_spatial_obs(self):
         self._validate_spatial_tissue_position("array_row", 0, 77)
         self._validate_spatial_tissue_position("in_tissue", 0, 1)
 
+        self._validate_spatial_is_primary_data()
+
+    def _validate_spatial_is_primary_data(self):
+        """
+        Validate is_primary_data for spatial datasets.
+        """
+        is_single = (
```
Comment (MillenniumFalconMechanic) [inline cellxgene_schema_cli/cellxgene_schema/validate.py:1391]: Could we use `_is_single()` here once this is rebased on `main`?

Comment (nayib-jose-gloria) [reply cellxgene_schema_cli/cellxgene_schema/validate.py:1391]: ah yes good call

File: cellxgene_schema_cli/cellxgene_schema/validate.py [hunk 8]
```diff
@@ -1382,6 +1382,25 @@ def _check_spatial_obs(self):
...
+            self.adata.uns["spatial"]["is_single"]
+            if "spatial" in self.adata.uns and "is_single" in self.adata.uns["spatial"]
+            else None
+        )
+        obs = getattr_anndata(self.adata, "obs")
+        if obs is None or "is_primary_data" not in obs:
+            return
+        if is_single is False and obs["is_primary_data"].any():
+            self.errors.append(
+                "When uns['spatial']['is_single'] is False, " "obs['is_primary_data'] must be False for all rows."
```
Comment (MillenniumFalconMechanic) [inline cellxgene_schema_cli/cellxgene_schema/validate.py:1401]: There's a break in the string here (`is False, " "obs['is_primary_data']`).

**Review by nayib-jose-gloria — COMMENTED** [review]
Note: I decided not to explicitly add a test_schema_compliance test case for allowing is_primary_data to be True or False if uns["spatial"]["is_single"] is None, because we implicitly cover this in `test_validate.py` by running `test__validate_with_h5ad_valid_and_labels`, which uses a fixture that has is_primary_data values True and False and no spatial in uns. I could be convinced to add the test case for clarity, but leaning towards no. LMK what you think

File: cellxgene_schema_cli/tests/test_schema_compliance.py [hunk 9]
```diff
@@ -1993,30 +2010,18 @@ def test_obsm_values_no_X_embedding__non_spatial_dataset(self, validator_with_ad
...
-        validator.adata.obs["assay_ontology_term_id"] = assay_ontology_term_id
-        validator.adata.obs["suspension_type"] = "na"
-        validator.adata.obs.loc[:, ["suspension_type"]] = validator.adata.obs.astype("category")
         validator.validate_adata()
         assert validator.errors == []
         assert validator.is_spatial is True
 
+    def test_obsm_values_no_X_embedding__slide_seq_v2_dataset(self, validator_with_slide_seq_v2_assay):
+        validator = validator_with_slide_seq_v2_assay
+        validator.adata.uns["default_embedding"] = "spatial"
```
Comment (MillenniumFalconMechanic) [inline cellxgene_schema_cli/tests/test_schema_compliance.py:2023]: Looks like this test is missing `validator.validate_adata()` and the corresponding assertions.

File: cellxgene_schema_cli/cellxgene_schema/validate.py [hunk 10]
```diff
@@ -1382,6 +1382,20 @@ def _check_spatial_obs(self):
...
+        self._validate_spatial_is_primary_data()
+
+    def _validate_spatial_is_primary_data(self):
+        """
+        Validate is_primary_data for spatial datasets.
+        """
+        obs = getattr_anndata(self.adata, "obs")
+        if obs is None or "is_primary_data" not in obs:
+            return
+        if self._is_single() is False and obs["is_primary_data"].any():
```
Comment (MillenniumFalconMechanic) [inline cellxgene_schema_cli/cellxgene_schema/validate.py:1394]: How are "secondary errors" typically handled? For example, should this error be reported for a non-spatial dataset that has (incorrectly) specified `uns.spatial.is_single = False`?

**Comment by MillenniumFalconMechanic** [comment]
Looks good, thanks Nayib! I have added a couple of comments.

File: cellxgene_schema_cli/tests/test_schema_compliance.py [hunk 11]
```diff
@@ -1993,30 +2010,18 @@ def test_obsm_values_no_X_embedding__non_spatial_dataset(self, validator_with_ad
...
-        self, validator_with_adata, assay_ontology_term_id, adata_spatial
-    ):
-        validator = validator_with_adata
-        validator.adata.obsm["harmony"] = validator.adata.obsm["X_umap"]
-        validator.adata.uns = adata_spatial.uns
-        validator.adata.uns["default_embedding"] = "harmony"
-        validator.adata.obsm["spatial"] = validator.adata.obsm["X_umap"]
+    def test_obsm_values_no_X_embedding__visium_dataset(self, validator_with_visium_assay):
+        validator = validator_with_visium_assay
+        validator.adata.uns["default_embedding"] = "spatial"
```
Comment (MillenniumFalconMechanic) [inline cellxgene_schema_cli/tests/test_schema_compliance.py:2015]: I am not familiar with the complete, required shape for `adata` here but I see there are a couple of differences between the previous version and `validator_with_visium_assay` (e.g. `validator.adata.obsm["harmony"]` is no longer defined, `validator.adata.uns["default_embedding"]` is now `spatial`). Are these updates expected/OK?

File: cellxgene_schema_cli/tests/test_schema_compliance.py (hunk 9 above)
Comment (nayib-jose-gloria) [reply cellxgene_schema_cli/tests/test_schema_compliance.py:2023]: sad, it must've gotten lost along the way during a merge resolution

File: cellxgene_schema_cli/cellxgene_schema/validate.py (hunk 10 above)
Comment (nayib-jose-gloria) [reply cellxgene_schema_cli/cellxgene_schema/validate.py:1394]: since we accumulate all errors by appending to "self.errors", something like that would be handled wherever the uns check is being done. In this case, we should handle that in check_spatial_uns. I'll double check if we already are and add it in if not.

File: cellxgene_schema_cli/tests/test_schema_compliance.py (hunk 11 above)
Comment (nayib-jose-gloria) [reply cellxgene_schema_cli/tests/test_schema_compliance.py:2015]: yes, its expected

for non-spatial datasets, not having an X-prefix embedding like `X_umap` should result in an error (see `test_obsm_values_no_X_embedding__non_spatial_dataset`). 

However, for spatial datasets, we do not require an X-prefix embedding, just 'spatial' embedding. This test case was originally modeled after `test_obsm_values_no_X_embedding__non_spatial_dataset`, which used 'harmony' as its non-X-prefix embedding to trigger the error. But I realized spatial is already a non-X-prefix embedding so we didn't need to add another one to test.

File: cellxgene_schema_cli/cellxgene_schema/validate.py (hunk 10 above)
Comment (nayib-jose-gloria) [reply cellxgene_schema_cli/cellxgene_schema/validate.py:1394]: Oh wait I see what you mean--yes, we do report secondary errors like this. That way, if the problem was that they did mean for it to be spatial but accidentally used the wrong assay, we still tell them in 1 run all the other issues associated with it. Generally, we try to report every possible error in one go even if they'd all be fixed with 1 upstream correction. With some exceptions--we don't validate the X matrix, for instance, if there are issues with dependent columns before then, because its an expensive operation to run if we know there will be failures. But we do print a warning letting them know that this was the case--`WARNING: "Validation of raw layer was not performed due to current errors, try again after fixing current errors."`

Comment (nayib-jose-gloria) [reply cellxgene_schema_cli/cellxgene_schema/validate.py:1394]: I do think its fine that we exit early w/o checking spatial values in `check_spatial_uns` if assay is not 'EFO:0010961' (Visium Spatial Gene Expression) or 'EFO:0030062' (Slide-seqV2), since curators have not requested otherwise and the errors would end up being the same for each one (i.e. `this key should not exist if its not visium` X times)

Comment (MillenniumFalconMechanic) [reply cellxgene_schema_cli/cellxgene_schema/validate.py:1398]: Got it, thanks Nayib!

**Review by MillenniumFalconMechanic — APPROVED** [review]
LGTM, thanks Nayib!

**Comment by nayib-jose-gloria** [comment]
@brian-mott ready for your review!

**Comment by brian-mott** [comment]
Looks good, thanks Nayib!
